### PR TITLE
Refactored Ingredients, implemented DepleteIngredient

### PR DESCRIPTION
**Pull request by hiyingnn**

### CODE REVIEW

**Comment by hiyingnn** [comment]
Currently there are GUI tests that are failing, will resolve by tonight.
Will implement tests for DeleteIngredient and DepleteIngredient by Friday.

File: src/main/java/seedu/address/model/ingredient/IngredientQuantity.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import seedu.address.commons.util.StringUtil;
+
+/**
+ * A class to represent the ingredientQuantity in an ingredient
+ */
+
+public class IngredientQuantity {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Ingredient's quantity should be non-zero unsigned integer.";
```
Comment (lingyanhao) [inline src/main/java/seedu/address/model/ingredient/IngredientQuantity.java:12]: still integer? i remember your user guide wanted to allow any decimal amount.

whichever approach you choose, ensure that implementation is consistent with user guide.

File: src/main/java/seedu/address/model/ingredient/IngredientQuantity.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        return ingredientQuantity;
+    }
+
+    /**
+     * Returns true if ingredientQuantity is valid.
+     * @param test
+     * @return
+     */
+    public static boolean isValidIngredientQuantity(String test) {
+        if (StringUtil.isNonZeroUnsignedInteger(test)) {
```
Comment (lingyanhao) [inline src/main/java/seedu/address/model/ingredient/IngredientQuantity.java:35]: do you want to allow 0 as a valid quantity?
like for example, i might run out of stock

File: src/main/java/seedu/address/model/ingredient/IngredientWarningAmount.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+
+    @Override
+    public boolean equals(Object other) {
+        boolean isitEquals = other == this // short circuit if same object
+            || (other instanceof IngredientWarningAmount// instanceof handles nulls
+            && ingredientWarningAmount == ((IngredientWarningAmount) other).getWarningAmount()); // state check
+        return isitEquals;
```
Comment (lingyanhao) [inline src/main/java/seedu/address/model/ingredient/IngredientWarningAmount.java:50]: why this extra variable? did you use for debugging? if so get rid of it

File: src/main/java/seedu/address/model/ingredient/IngredientQuantity.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public int getQuantity() {
+        return ingredientQuantity;
+    }
+
+    /**
+     * Returns true if ingredientQuantity is valid.
+     * @param test
+     * @return
```
Comment (lingyanhao) [inline src/main/java/seedu/address/model/ingredient/IngredientQuantity.java:32]: javadoc: don't have a blank '@return'. if you dont want to put anything, leave it blank

File: src/main/java/seedu/address/model/ingredient/IngredientName.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    // Identity fields
+    private final String ingredientName;
+
+    /**
+     * Constructs a {@code IngredientName}.
+     *
+     * @param name A valid ingredient name corresponding to VALIDATION_REGEX
+     */
+    public IngredientName(String name) {
+        this.ingredientName = name;
```
Comment (lingyanhao) [inline src/main/java/seedu/address/model/ingredient/IngredientName.java:23]: should exercise some defensive programming here- what happens if someone calls your constructor without checking the regex? 

might want to throw exception (this applies to the other parts as well- i'm only writing this comment once).

File: src/main/java/seedu/address/logic/commands/DepleteIngredientCommand.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+import seedu.address.model.Model;
+import seedu.address.model.ingredient.Ingredient;
+import seedu.address.model.ingredient.IngredientQuantity;
+import seedu.address.model.person.exceptions.DuplicateItemException;
+
+/**
+ * Depletes the quantity of existing ingredient in the restaurant book.
+ */
+
+public class DepleteIngredientCommand extends Command {
```
Comment (lingyanhao) [inline src/main/java/seedu/address/logic/commands/DepleteIngredientCommand.java:23]: deplete and delete look very similar- consider having a different name for the command to be less confusing for the reader?

File: src/main/java/seedu/address/logic/commands/DepleteIngredientCommand.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     * New Ingredient with depleted quantity is returned.
+     * @param ingredientToDeplete
+     * @return Ingredient
+     */
+    private static Ingredient createDepletedIngredient(Ingredient ingredientToDeplete, IngredientQuantity quantity) {
+        assert ingredientToDeplete != null;
+
+        int currentQuantity = ingredientToDeplete.getIngredientQuantity().getQuantity();
+        int newQuantity = currentQuantity - quantity.getQuantity();
```
Comment (lingyanhao) [inline src/main/java/seedu/address/logic/commands/DepleteIngredientCommand.java:88]: What happens if `newQuantity` is negative?

File: src/main/java/seedu/address/logic/commands/DepleteIngredientCommand.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            throw new CommandException(Messages.MESSAGE_INVALID_INGREDIENT_DISPLAYED_INDEX);
+        }
+
+        Ingredient ingredientToDeplete = lastShownList.get(index.getZeroBased());
+        Ingredient depletedIngredient = createDepletedIngredient(ingredientToDeplete, quantityToDeplete);
+
+        try {
+            model.setIngredient(ingredientToDeplete, depletedIngredient);
+        } catch (DuplicateItemException e) {
+            throw new CommandException(MESSAGE_DUPLICATE);
```
Comment (lingyanhao) [inline src/main/java/seedu/address/logic/commands/DepleteIngredientCommand.java:71]: will depleting an ingredient ever cause a duplicate?
(i.e. can you construct a test case such that this exception is thrown)

File: src/main/java/seedu/address/logic/commands/DepleteIngredientCommand.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            + "Input value will be added to existing value.\n"
+            + "Parameters: "
+            + PREFIX_INDEX + "INDEX "
+            + PREFIX_INGREDIENT_QUANTITY + "QUANTITY "
+            + "Example: " + COMMAND_WORD + " "
+            + PREFIX_INDEX + "1 "
+            + PREFIX_INGREDIENT_QUANTITY + "10 ";
+
+    public static final String MESSAGE_SUCCESS = "Depleted Ingredient: %1$s";
+    public static final String MESSAGE_DUPLICATE = "Ingredient is already in list";
```
Comment (lingyanhao) [inline src/main/java/seedu/address/logic/commands/DepleteIngredientCommand.java]: I don't think there will ever be duplicates (it doesn't make sense)
However, you need a different kind of error message- if I deplete an amount greater than what I currently have, an error message should be thrown.

**Pull request merged**